pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -193,29 +287,156 @@ contract InvestmentPool is BaseWeightedPool, ReentrancyGuard {
+        _require(getSwapEnabled(), Errors.SWAPS_DISABLED);
+
+        return super._onSwapGivenIn(swapRequest, currentBalanceTokenIn, currentBalanceTokenOut);
+    }
+
+    function _onSwapGivenOut(
+        SwapRequest memory swapRequest,
+        uint256 currentBalanceTokenIn,
+        uint256 currentBalanceTokenOut
+    ) internal view override returns (uint256) {
+        _require(getSwapEnabled(), Errors.SWAPS_DISABLED);
+
+        return super._onSwapGivenOut(swapRequest, currentBalanceTokenIn, currentBalanceTokenOut);
+    }
+
+    /**
+     * @dev Extend ownerOnly functions to include the LBP control functions
+     */
+    function _isOwnerOnlyAction(bytes32 actionId) internal view override returns (bool) {
+        return
+            (actionId == getActionId(InvestmentPool.setSwapEnabled.selector)) ||
+            (actionId == getActionId(InvestmentPool.updateWeightsGradually.selector)) ||
+            super._isOwnerOnlyAction(actionId);
+    }
+
+    function _onJoinPool(
+        bytes32 poolId,
+        address sender,
+        address recipient,
+        uint256[] memory balances,
+        uint256 lastChangeBlock,
+        uint256 protocolSwapFeePercentage,
+        uint256[] memory scalingFactors,
+        bytes memory userData
+    )
+        internal
+        override
+        returns (
+            uint256 bptAmountOut,
+            uint256[] memory amountsIn,
+            uint256[] memory dueProtocolFeeAmounts
+        )
+    {
+        // TokenInForExactBPTOut is not proportional - so fail immediately if swaps are disabled
+        _require(
+            getSwapEnabled() || userData.joinKind() != JoinKind.TOKEN_IN_FOR_EXACT_BPT_OUT,
+            Errors.DISPROPORTIONATE_JOIN_OR_EXIT
+        );
+
+        (bptAmountOut, amountsIn, dueProtocolFeeAmounts) = super._onJoinPool(
+            poolId,
+            sender,
+            recipient,
+            balances,
+            lastChangeBlock,
+            protocolSwapFeePercentage,
+            scalingFactors,
+            userData
+        );
+
+        // If the pool is paused, only allow proportional join
+        if (!getSwapEnabled()) {
+            // We know the bptOut from the join - so proportional exit with that bptOut should reproduce amountsIn,
+            // if the join was proportional
+            uint256[] memory amountsOut = WeightedMath._calcTokensOutGivenExactBptIn(
+                balances,
+                bptAmountOut,
+                totalSupply().add(bptAmountOut)
+            );
+            uint256 diff;
+
+            for (uint256 i = 0; i < amountsOut.length; i++) {
+                diff = amountsOut[i] > amountsIn[i] ? amountsOut[i].sub(amountsIn[i]) : amountsIn[i].sub(amountsOut[i]);
+
+                _require(
+                    amountsIn[i] > 0 && diff.divUp(amountsIn[i]) <= _PROPORTIONAL_TOLERANCE_PCT,
+                    Errors.DISPROPORTIONATE_JOIN_OR_EXIT
+                );
+            }
```

Comment:
While I think this is just about the best we can do with what we currently have, I can't help but wonder if a better solution might be to add a new `JoinKind` to `BaseWeightedPool` that is strictly proportional. Something like `TOKENS_IN_FOR_EXACT_BPT_OUT` (note the 's' in 'tokens'). But I'm lacking some context on why this was originally omitted (whereas the equivalent `ExitKind` exists). Maybe @nventuro could shed some light?

The way I see it, no matter how low we set this percentage threshold, it could always be gamed for a sufficiently large pool. As long as the arbitrageur has reserves of all tokens, they could continuously join/exit to make swaps on one pair. And that would completely ruin the value proposition of the paused state. If we could instead strictly enforce proportionality, then the integrity of the paused state is maintained without exceptions.

Reply:
Yes - I thought about that too, but hesitated to add a whole new join kind. Makes sense; let's ask.

Reply:
After chatting with @nventuro, we both concur that adding the new `JoinKind` makes the most sense here. It sounds like it was only omitted to begin with because it seemed redundant at the time - in the absence of a pause, there is actually no good UX reason to support this type of join over an uneven join (which can incidentally be proportional). But now that we have the pause capability, I think adding this strictly enforced proportional join to `BaseWeightedPool` is a necessity, and there shouldn't be any blockers to doing this. @EndymionJkb, are you content to insert that change into this PR in lieu of the functionality above?

Reply:
Yes - that's much cleaner. Will do.